### PR TITLE
Add multiplayer server disconnection flow

### PR DESCRIPTION
**Pull request by peppy**
This handles multiple cases of the client getting stuck (or worse, crashing) when the multiplayer server connection is lost. It also provides better messaging and flow when attempting to create a room while the server is unavailable.

### CODE REVIEW

**Comment by bdach** [comment]
Conflict is trivial in resolution (need to take both changes) but would like to discuss the above before proceeding.

File: osu.Game/Online/RealtimeMultiplayer/RealtimeMultiplayerClient.cs [hunk 1]
```diff
@@ -130,7 +130,11 @@ protected override Task<MultiplayerRoom> JoinRoom(long roomId)
         public override async Task LeaveRoom()
         {
             if (!isConnected.Value)
+            {
+                // even if not connected, make sure the local room state can be cleaned up.
+                await base.LeaveRoom();
                 return;
+            }
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Online/RealtimeMultiplayer/RealtimeMultiplayerClient.cs:137]: that sounds fine too. i usually just prefer an explicit exit condition at the start of the method.

File: osu.Game/Screens/Multi/RealtimeMultiplayer/RealtimePlayer.cs [hunk 2]
```diff
@@ -43,17 +45,26 @@ private void load()
...
             if (!startedEvent.Wait(TimeSpan.FromSeconds(30)))
             {
                 Logger.Log("Failed to start the multiplayer match in time.", LoggingTarget.Runtime, LogLevel.Important);
 
-                Schedule(() =>
-                {
-                    ValidForResume = false;
-                    this.Exit();
-                });
+                Schedule(PerformImmediateExit);
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Multi/RealtimeMultiplayer/RealtimePlayer.cs:74]: I was going based off method name; maybe `this.Exit` is more suited here, will double check.  Maybe `PerformImmediateExit` should be renamed in that case?

[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Screens/Multi/RealtimeMultiplayer/RealtimePlayer.cs:74]: Yeah, definitely.

File: osu.Game/Online/RealtimeMultiplayer/RealtimeMultiplayerClient.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Online/RealtimeMultiplayer/RealtimeMultiplayerClient.cs:137]: On another look I can be convinced either way. The way it is right now is less subtle than what I proposed, so the current way is probably better.

**Comment by bdach** [comment]
I've pushed in the change to `Player` discussed on discord, seemed to work OK in testing. Did not move the current screen to start of method, as I checked blame and saw https://github.com/ppy/osu/pull/5154, which had it there already once. The test added in that pull didn't fail when I tried moving, but I'd rather be safe than sorry.

**Review by bdach — APPROVED** [review]

**Pull request merged**